src/clj/tic_tac_toe_clojure/core.clj

```diff
@@ -49,47 +50,78 @@
    (symbols-at board [2 4 6])])
 
 (defn join-sections [board]
-  (concat (get-rows board) (concat (get-columns board)) (concat(get-diagonals board))))
+  (concat (get-rows board) (concat (get-columns board)) (concat (get-diagonals board))))
 
-(defn symbols-equal? [section]
-  (and (= (nth section 0) (nth section 1) (nth section 2)) (not= (apply str section) "___")))
+(defn symbols-equal? [section symbol]
+  (and (= (nth section 0) (nth section 1) (nth section 2) symbol) (not= (apply str section) "___")))
 
-(defn three-aligned? [board]
-  (some true? (for [section (join-sections board)] (symbols-equal? section))))
+(defn three-aligned? [board symbol]
+  (some true? (for [section (join-sections board)] (symbols-equal? section symbol))))
 
 (defn board-full? [board]
   (if (some #{"_"} board) false true))
 
 (def end-game "Game is over")
 
-(defn game-over? [board]
-  (if (or (three-aligned? board) (board-full? board)) true false))
+(defn game-over? [board symbol1 symbol2]
+  (or (three-aligned? board symbol1) (three-aligned? board symbol2) (board-full? board)))
+
+(defn get-empty-spots [board]
+  (remove #{"_"} (flatten (find-empty-spots board))))
+
+(defn evaluate-board [board current-player opponent]
+  (cond
+    (three-aligned? board current-player) 10
+    (three-aligned? board opponent) -10
+    :else 0))
+
+(defn minimax [board maximising-player minimising-player]
+    (if (game-over? board maximising-player minimising-player)
+      (evaluate-board board maximising-player minimising-player)
+      (* -1 (val (apply max-key val (scored-moves board minimising-player maximising-player))))))
```

Comment:
Technically, I think this is actually negamax - not minimax. This is because both players are always finding the maximally-scoring move - it's just that this score is being 'flipped' (multiplied by -1) at each level of depth. If it were minimax, you'd have a minimising function, where the minimum score is chosen on the opponent's turn.
Also, at this line 81, is there a way to re-use 'choose-best-move' recursively here instead of the nested function calls you've used? I am not sure if this is possible as one uses '(val' and one uses '(key' but perhaps something to think about.

Reply:
@devlinglasman - I'm gonna leave it as is, because I see you point about using `choose-best-move` recursively in order to keep it DRY, but ultimately, while they look similar they do different things (one returns a score, the other a position). 

I think extracting `(val (apply max-key val ...` at L81 out into a `choose-best-score` function would make more sense than using one function to do two different things if I was really concerned about that level of granular readability, but I don't really want a `choose-highest-value` function that cognitively overloads the reader some more by forcing them to figure out what that value is within other functions.